library/core/src/fmt/mod.rs

```diff
@@ -265,6 +265,26 @@ pub struct ArgumentV1<'a> {
     formatter: fn(&Opaque, &mut Formatter<'_>) -> Result,
 }
 
+/// This struct represents the unsafety of constructing an `Arguments`.
+/// It exists, rather than an unsafe function, in order to simplify the expansion
+/// of `format_args!(..)` and reduce the scope of the `unsafe` block.
+#[allow(missing_debug_implementations)]
+#[doc(hidden)]
+#[non_exhaustive]
+#[unstable(feature = "fmt_internals", reason = "internal to format_args!", issue = "none")]
+pub struct UnsafeArg;
```

Comment:
Since the struct is public, it can be constructed directly without an unsafe block.

Reply:
It's non_exhaustive so I don't think that's true.

Reply:
`non_exhaustive` is insufficient, it only applies across crate boundaries. This compiles:
```rust
mod test {
#[non_exhaustive]
pub struct UnsafeArg;
}

fn main() {
    let x = test::UnsafeArg;
}
```
So I think this really should have a private field.

Reply:
Ah, good catch. Yeah, would be good to add a private field then.

Reply:
Like this? `struct UnsafeArg(())`?

Reply:
That should work. We can probably drop non_exhaustive at that point as well.

Reply:
> `non_exhaustive` is insufficient, it only applies across crate boundaries.

Isn't that enough since `UnsafeArg` is in the core crate?

Reply:
We shouldn't rely on that - it doesn't make this unsound for external users, but libcore itself can still get it wrong.

Reply:
I see.